vdr/vdr.go

```diff
@@ -0,0 +1,168 @@
+// It is also a Runnable, Diagnosable and Configurable Nuts Engine.
+type VDR struct {
+	Config            Config
+	store             types.Store
+	network           networkPkg.NetworkClient
+	OnChange          func(registry *VDR)
+	networkAmbassador network.Ambassador
+	configOnce        sync.Once
+	_logger           *logrus.Entry
+	closers           []chan struct{}
+	didDocCreator     types.DocCreator
+}
+
+var instance *VDR
+var oneVDR sync.Once
+
+// Instance returns the singleton VDR
+func Instance() *VDR {
+	if instance != nil {
+		return instance
+	}
+	oneVDR.Do(func() {
+		instance = NewVDR(DefaultConfig(), crypto.Instance(), networkClient.NewNetworkClient())
+	})
+
+	return instance
+}
+
+// NewVDR creates a new VDR with provided params
+func NewVDR(config Config, cryptoClient crypto.KeyStore, networkClient pkg.NetworkClient) *VDR {
+	return &VDR{
+		Config:        config,
+		network:       networkClient,
+		_logger:       logging.Log(),
+		store:         store.NewMemoryStore(),
+		didDocCreator: NutsDocCreator{keyCreator: cryptoClient},
+	}
+}
+
+// Configure initializes the db, but only when in server mode
+func (r *VDR) Configure() error {
+	var err error
+
+	r.configOnce.Do(func() {
+		cfg := core.NutsConfig()
+		if cfg.Mode() == core.ServerEngineMode {
+			if r.networkAmbassador == nil {
+				r.networkAmbassador = network.NewAmbassador(r.network)
+			}
+		}
+	})
+	return err
+}
+
+// Start initiates the routines for auto-updating the data
+func (r *VDR) Start() error {
+	if core.NutsConfig().Mode() == core.ServerEngineMode {
+		r.networkAmbassador.Start()
+	}
+	return nil
+}
+
+// Shutdown cleans up any leftover go routines
+func (r *VDR) Shutdown() error {
+	if core.NutsConfig().Mode() == core.ServerEngineMode {
+		logging.Log().Debug("Sending close signal to all routines")
+		for _, ch := range r.closers {
+			ch <- struct{}{}
+		}
+		logging.Log().Info("All routines closed")
+	}
+	return nil
+}
+
+// Diagnostics returns the diagnostics for this engine
+func (r *VDR) Diagnostics() []core.DiagnosticResult {
+	return []core.DiagnosticResult{}
+}
+
+func (r *VDR) getEventsDir() string {
+	return r.Config.Datadir + "/events"
+}
+
+// Create generates a new DID Document
+func (r VDR) Create() (*did.Document, error) {
+	doc, err := r.didDocCreator.Create()
+	if err != nil {
+		return nil, fmt.Errorf("could not create did document: %w", err)
+	}
+
+	// Fixme: The doc should not be stored but send to the network.
+	metaData := types.DocumentMetadata{
+		Created: time.Now(),
+		Version: 0,
+	}
+	err = r.store.Write(*doc, metaData)
+	if err != nil {
+		return nil, fmt.Errorf("could not store created did document: %w", err)
+	}
+	return doc, nil
+}
+
+// Resolve resolves a DID Document based on the DID.
+func (r VDR) Resolve(dID did.DID, metadata *types.ResolveMetaData) (*did.Document, *types.DocumentMetadata, error) {
```

Comment:
proper capitalization `ResolveMetaData`; in other places we call MetaData just metadata (or Metadata)